python/amici/sbml_import.py

```diff
@@ -742,16 +763,52 @@ def _process_rules(self) -> None:
-                self.stoichiometric_matrix = \
-                    self.stoichiometric_matrix.subs(variable, formula)
+                if variable not in self.compartment_symbols: # change to only self.compartment_rules symbols?
+                    self.stoichiometric_matrix = \
+                        self.stoichiometric_matrix.subs(variable, formula)
+                #else:
+                #    self.stoichiometric_matrix = \
+                #        self.stoichiometric_matrix.subs(variable, self.compartment_volume[list(self.compartment_symbols).index(variable)])
 
             if variable in specvars:
                 raise SBMLException('Species assignment rules are currently'
                                     ' not supported!')
 
             if variable in compartmentvars:
-                raise SBMLException('Compartment assignment rules are'
-                                    ' currently not supported!')
+                if rule.getTypeCode() == sbml.SBML_ASSIGNMENT_RULE:
+                    self.compartment_rules[variable] = {
+                            'type_code': sbml.SBML_ASSIGNMENT_RULE,
+                            'formula': formula,
+                            # setting v0 might be unnecessary. currently in place to satisfy the ODEModel _equation_from_component method (which I edited to require v0...)
+                            'v0': self.compartment_volume[list(self.compartment_symbols).index(variable)].subs(
+                                self.symbols['species']['identifier'],
+                                self.symbols['species']['value']
+                            )
+                        }
+                    self.compartment_volume[list(self.compartment_symbols).index(variable)] = formula
+                    assignments[str(variable)] = formula
+                    #continue # avoid simplifying substitution of the compartment expression, to avoid possibly circular substitution
+                elif rule.getTypeCode() == sbml.SBML_RATE_RULE:
+                    self.compartment_rules[variable] = {
+                            'type_code': sbml.SBML_RATE_RULE,
+                            'formula': formula,
+                            'v0': self.compartment_volume[list(self.compartment_symbols).index(variable)]
+                        }
+
+                    self.symbols['species']['identifier'] = self.symbols['species']['identifier'].col_join(sp.Matrix([variable]))
+                    self.symbols['species']['name'].append('')
+                    self.symbols['species']['value'] = self.symbols['species']['value'].col_join(sp.Matrix([variable]))
+                    #self.symbols['species']['value'] = self.symbols['species']['value'].col_join(sp.Matrix([self.compartment_rules[variable]['v0']]))
+                    self.species_index.update({sp.sstr(variable): len(self.species_index)}) #necessary?
+                    self.stoichiometric_matrix = self.stoichiometric_matrix.col_join(
```

Comment:
Extending like this is likely to be inefficient for larger matrices and the stoichiometric matrix can easily have multiple million entries. By moving the above code to `_process_species` this allocation is already partially adressed. I think it would be great to split this code in two, one that precomputes most of what is done here, but doesn't update `stoichiometric_matrix` and `flux_vector` yet. `_process_reactions()` could then correctly initialize the stoichiometric matrix based on the precomputed variables and this function would just update `stoichiometric_matrix` and `flux_vector` accordingly.

Reply:
The implementation of rate rules no longer involves `stoichiometric_matrix`/`flux_vector`.